main/design_documents/ARM_NEON_SVE_BRIDGE.rst

```diff
@@ -0,0 +1,56 @@
+=======================================
+Design Document for ARM_NEON_SVE_BRIDGE
+=======================================
+
+This document describes the design of the intrinsics defined in the ("NEON_SVE bridge")
+<../acle.rst#neon-sve-bridge>. These intrinsics add a
+convienient way to use NEON and SVE vectors interchangeably in C code with
+no performance overhead.
+
+Location
+########
+
+The intrinsics will be in a new header file called ``arm_extensions_bridge.h``.
+
+This header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.
+
+The intrinsics will be conditionally defined with
+
+.. code:: c
+
+    #if !defined(__ARM_FEATURE_NEON) || !defined(__ARM_FEATURE_SVE)
+        #error "arm_extensions_bridge.h requires -march=+simd+sve"
```

Comment:
Not sure this is the right error to give here. Trying to use this option with GCC gives `cc1: error: missing arch name in '-march=+simd+sve'`. Clang gives `clang-13: error: the clang compiler does not support '-march=+simd+sve'`.

Reply:
`requires -march=armv8-a+simd+sve`?

Reply:
If Neon support is mandatory in AArch64 I wonder if theres any point in me asserting against `__ARM_FEATURE_NEON` anywhere at all

Reply:
```
int main(){
#ifdef __ARM_FEATURE_NEON
        printf("neon enabled\n");
#endif
#ifdef __ARM_FEATURE_SVE
        printf("sve enabled\n");
#endif
        return 0;
}
```
Compiling the above with the trunk of clang with `-march=armv8-a+simd+sve` only outputs `sve enabled`... interesting

Reply:
I'm guessing i should just be using `__ARM_NEON` then?

Reply:
> If Neon support is mandatory in AArch64 I wonder if theres any point in me asserting against `__ARM_FEATURE_NEON` anywhere at all

Good thought - it's difficult to imagine a scenario where the NEON feature is disabled but SVE is enabled, so I believe testing for SVE should be sufficient.

Reply:
There's still the problem that SVE doesn't bring in neon types and we still need to be sure we have the neon types defined when we include this file

Reply:
I thought the intent was to resolve that issue by forward declaring the neon types?

Reply:
https://github.com/ARM-software/acle/pull/72#discussion_r740912334 we mentioned it here but didnt explicitly agree on it. From that conversation I was under the impression we were checking the `__ARM_FEATURE` macros to ensure the types had already been defined

Reply:
Ah, last I saw (and we discussed), you said 'forward declaring the types is probably the best bet' which sounded good to me, since it makes the order of includes not matter, which is important in the face of tools which might reorder the headers and helps with the principle of least surprise for users. I'll replied there too.